### ISSUE TITLE
Args objects

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#131]
I ran into a slight usability problem when doing cross-language APIs (written in JavaScript, consumed in Python).  Below is a brief description and one a possible approach to a solution.

The idiomatic JavaScript way of offering "named parameters" is to use a parameter object.  This is especially useful for Coconut resource types, where the number of parameters available is often quite large, especially given that most of them are optional.  For example:

    let grp = new ec2.SecurityGroup("web-secgrp", { description: "Enable HTTP access" });

The idiomatic Python way of doing this, OTOH, is to use real named parameters:

    grp = ec2.SecurityGroup("web-secgrp", description="Enable HTTP access")

Because our AWS resources are written in JavaScript and consumed in Python, however, the Python code needed to create a new resource doesn't look so idiomatic:

    grp = ec2.SecurityGroup("web-secgrp", ec2.SecurityGroupArgs(description="Enable HTTP access"))

This probably isn't a dealbreaker, but it's unidiomatic enough that Python developers will almost certainly regularly forget to create the args object, and then receive annoying runtime failures.

One idea on how to solve this would be to let parameter types in the JavaScript side derive from a base class, and then do some patchwork at the seams.  For instance, perhaps `SecurityGroupArgs` derives from a base class, `coconut.Args`, and then we do magic similar to C# params arguments.

In other words, a sequence of Python-esque named arguments in a parameter slot expecting a `coconut.Args` object auto-expands to the creation of the parameter object.  The original Python sample above therefore automatically expands into the second at runtime.

Unfortunately, it's not yet clear to me how to make the reverse work.  Perhaps we should eliminate named/optional parameters altogether from CocoIL and simply transform them into idiomatic `coconut.Args`-style parameter passing under the hood.  This seems to work in both directions.
